src/analysis/processing/qgsalgorithmjoinbylocation.cpp

```diff
@@ -0,0 +1,350 @@
+  int step = mBaseSource->featureCount() > 0 ? 100.0 / mBaseSource->featureCount() : 1;
+  QgsFeatureIterator joinIterator = mJoinSource->getFeatures();
+  int i = 0;
+  while ( joinIter.nextFeature( f ) )
+  {
+    if ( feedback->isCanceled() )
+      break;
+    feedback->setProgress( i * step );
+    if ( processFeatures( f, context, feedback ) )
+      joinedCount++;
+    i++;
+  }
+
+  if ( !mDiscardNonMatching )
+  {
+    QgsFeature f2;
+    QgsFeatureRequest remainings = QgsFeatureRequest().setFilterFids( mUnjoinedIds );
+    QgsFeatureIterator remainIter = mBaseSource->getFeatures( remainings );
+    while ( remainIter.nextFeature( f2 ) )
+    {
+      if ( mJoinedFeatures )
+      {
+        QgsAttributes attributes = f2.attributes();
+        attributes.append( mEmptyAttrs );
+        QgsFeature outputFeature( f2 );
+        outputFeature.setAttributes( attributes );
+        mJoinedFeatures->addFeature( outputFeature, QgsFeatureSink::FastInsert );
+      }
+      if ( mUnjoinedFeatures )
+        mUnjoinedFeatures->addFeature( f2, QgsFeatureSink::FastInsert );
+    }
+  }
+
+  QVariantMap outputs;
+  if ( mJoinedFeatures )
+  {
+    outputs.insert( QStringLiteral( "OUTPUT" ), mJoinedFeaturesId );
+  }
+  if ( mUnjoinedFeatures )
+  {
+    outputs.insert( QStringLiteral( "NON_MATCHING" ), mUnjoinedFeaturesId );
+  }
+  outputs.insert( QStringLiteral( "JOINED_COUNT" ), joinedCount );
+  return outputs;
+}
+
+bool QgsJoinByLocationAlgorithm::featureFilter( const QgsFeature &feature, QgsGeometryEngine *engine ) const
+{
+  const QgsAbstractGeometry *geom = feature.geometry().constGet();
+  bool ok = false;
+  for ( const auto  predicate : mPredicates )
+  {
+    switch ( predicate )
+    {
+      case 0:
+        if ( engine->intersects( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 1:
+        if ( engine->within( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 2:
+        if ( engine->isEqual( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 3:
+        if ( engine->touches( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 4:
+        if ( engine->overlaps( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 5:
+        if ( engine->contains( geom ) )
+        {
+          ok = true;
+        }
+        break;
+      case 6:
+        if ( engine->crosses( geom ) )
+        {
+          ok = true;
+        }
+        break;
+    }
+    if ( ok )
+      return ( ok );
+  }
+  return ( ok );
+}
+
+bool QgsJoinByLocationAlgorithm::processFeatures( QgsFeature &joinFeature, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+
+  if ( !joinFeature.hasGeometry() )
+    return false;
+  const QgsGeometry featGeom = joinFeature.geometry();
+  QgsGeometryEngine *engine = QgsGeometry::createGeometryEngine( featGeom.constGet() );
```

Comment:
```suggestion
  std::unique_ptr< QgsGeometryEngine > engine( QgsGeometry::createGeometryEngine( featGeom.constGet() ) );
```

but engine should be lazy constructed only on getting the first matching feature -- otherwise we are doing unnecessary work when in the case that there's no matching features based on the bounding box request alone

Reply:
I hate unique ptrs, they are so finnicky, I have no clue why I get the error that its deleted during the travis build.